src/adapter/src/coord/sequencer/inner/copy_from.rs

```diff
@@ -0,0 +1,181 @@
+        &mut self,
+        ctx: ExecuteContext,
+        plan: plan::CopyFromPlan,
+        target_cluster: TargetCluster,
+    ) {
+        let plan::CopyFromPlan {
+            id,
+            source,
+            columns: _,
+            params: _,
+        } = plan;
+
+        let from_expr = match source {
+            CopyFromSource::Url(from_expr) => from_expr,
+            CopyFromSource::Stdin => {
+                unreachable!("COPY FROM STDIN should be handled elsewhere")
+            }
+        };
+
+        let eval_url = |from: HirScalarExpr| -> Result<Url, AdapterError> {
+            let style = ExprPrepStyle::OneShot {
+                logical_time: EvalTime::NotAvailable,
+                session: ctx.session(),
+                catalog_state: self.catalog().state(),
+            };
+            let mut from = from.lower_uncorrelated()?;
+            prep_scalar_expr(&mut from, style)?;
+
+            let temp_storage = RowArena::new();
+            let eval_result = from.eval(&[], &temp_storage)?;
+
+            if eval_result == Datum::Null {
+                coord_bail!("COPY FROM target value cannot be NULL");
+            }
+            let eval_string = match eval_result {
+                Datum::Null => coord_bail!("COPY FROM target value cannot be NULL"),
+                Datum::String(url_str) => url_str,
+                other => coord_bail!("programming error! COPY FROM target cannot be {other}"),
+            };
+
+            Url::parse(eval_string)
+                .map_err(|err| AdapterError::Unstructured(anyhow::anyhow!("{err}")))
+        };
+        let url = return_if_err!(eval_url(from_expr), ctx);
+
+        let dest_table = self
+            .catalog()
+            .get_entry(&id)
+            .table()
+            .expect("TODO SHOULD BE A TABLE");
+
+        let collection_id = dest_table.global_id_writes();
+        let (_, ingestion_id) = self.transient_id_gen.allocate_id();
+        let request = OneshotIngestionRequest {
+            source: mz_storage_types::oneshot_sources::ContentSource::Http { url },
+            format: mz_storage_types::oneshot_sources::ContentFormat::Csv,
+        };
+
+        let cluster_id = self
+            .catalog()
+            .resolve_target_cluster(target_cluster, ctx.session())
+            .expect("TODO do this in planning")
+            .id;
+
+        // When we finish staging the Batches in Persist, we'll send a command
+        // to the Coordinator.
+        let command_tx = self.internal_cmd_tx.clone();
+        let conn_id = ctx.session().conn_id().clone();
+        let closure = Box::new(move |batches| {
+            let _ = command_tx.send(crate::coord::Message::StagedBatches {
+                conn_id,
+                table_id: id,
+                batches,
+            });
+        });
+        // Stash the execute context so we can cancel the COPY.
+        self.active_copies
+            .insert(ctx.session().conn_id().clone(), ctx);
```

Comment:
I'm not sure if this has been discussed, but if we cancel after the batch has been staged, then will we leak the batch in persist?

Reply:
We would leak the batches. If we cancel the request we could spawn a task that will wait for the response and clean them up, but concurrently Persist is also working on a leaked blob detector so this shouldn't be too much of an issue